Maker decides revoke commit parameters based on proposed `commit-txid`

da-kami opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/2475

We reproduce the behavior in a test:

On `testnet`, when the maker is ahead with rollovers we run into error `Counterparty sent invalid revocation sk` when rolling over from a previous commit.
This was not detected by the `retry_rollover_an_open_cfd` test because, even though we trigger rolling over from a previous commit-txid, the taker's DLC actually matched the one of the maker. The test setup did no properly simulate that the maker is in fact one rollover ahead of the taker.

As it is hard to remove events, the easiest way to simulate the previous rollover on the taker side, is to add another `RolloverCompleted` event with the old data. This should be sufficient to simulate that the taker has in fact the old state and the maker is ahead.

---

For the fix the maker decides the revocation parameters based on the `commit-txid`, taking them either from the current DLC or from the revoked commits.

klochowicz commented on daemon-tests/src/lib.rs:
```diff
@@ -462,6 +485,30 @@ impl Taker {
+    ///
+    /// Note that the projection does not get updated, this change only manipulates the database!
+    /// When triggering another rollover this data will be loaded and used.
+    pub async fn simulate_previous_rollover(
```
`simulate_one_rollover_ahead` might be a better name?

luckysori replied:
I'm not sure about `simulate_one_rollover_ahead` because that depends on the `Dlc` that you pass in as an argument. The current name might suggest that we're overwriting the previous rollover, but we're actually just adding a new one. An alternative could be the straightforward `append_rollover`.

It's a nifty trick that we can just _append_ a rollover completed event and make it seem like the taker is actually _behind_. But we really want to eventually be able to run a test like this based on a state that is truly behind i.e. rollover failed instead of rollover completed.

da-kami replied:
Yeah, I'm not proud of the nifty trick, I just wanted to make sure to reproduce it and see the test turn green. 
In the long run I think it'd be better to test this by removing data rather than adding, but that's currently not easily possible.

klochowicz replied:
`append_rollover` sounds good too!

luckysori replied:
I think it's great that you found a clever workaround. Otherwise we wouldn't have a test!

klochowicz reviewed (commented):
kudos for reproducing the problem in a test!

bonomat reviewed (approved):
Good find 👏👏

luckysori commented on daemon-tests/src/lib.rs:
```diff
@@ -462,6 +485,30 @@ impl Taker {
+                id,
+                EventKind::RolloverCompleted {
+                    dlc: Some(dlc),
+                    // Funding fee irrelevant because only CompleteFee is used
+                    funding_fee: dummy_funding_fee(),
```
❓ Interesting, but why do we have a `funding_fee` field if it's irrelevant? Are we planning to remove this from this event?

luckysori commented on daemon-tests/tests/happy_path.rs:
```diff
@@ -523,6 +525,16 @@ async fn retry_rollover_an_open_cfd() {
         "The maker and taker should have the same commit_txid after the first rollover"
```
🔧 I would remove all these asserts on the taker's first rollover success because:

- They're irrelevant for the test. We actually want to immediately "forget" this rollover and pretend it never worked.
- They make the test harder to understand.

da-kami replied:
👍

luckysori reviewed (approved):
Great job! I'll make some minor tweaks and get this merged so that we can try to release sooner.

luckysori commented on daemon-tests/tests/happy_path.rs:
```diff
@@ -506,36 +508,29 @@ async fn retry_rollover_an_open_cfd() {
 
-    // We mock the initial oracle event-id for the second rollover again
+    // We mock the initial oracle event-id for the rollover retry again
     mock_oracle_announcements(
         &mut maker,
         &mut taker,
         contract_setup_oracle_data_announcement.clone(),
     )
     .await;
```
I would like to leave a better comment here as to why this is necessary. I feel like it's just a quirk of the test, but I might be missing something! cc @da-kami

da-kami replied:
To some extent it is a quirk, because we only have two events that we currently use in the test, i.e. we use `exmaple_0` for contract setup, then `example_1` for the first rollover, and `example_0` for the second one. I originally opted for this so I can assert that the value changes, i.e. the mocked event is correctly picked when rolling over from a previous commit. But re-using `example_0` is indeed confusing and should be changed.

bonomat commented on model/src/rollover.rs:
```diff
@@ -84,3 +99,167 @@ impl RolloverParams {
+        let revoked_commits = self
+            .revoked_commit
+            .iter()
+            .take_while(|commit| commit.txid != revoked_commit_txid)
```
🔧
Is it guaranteed that `revoked_commit` is sorted?
It's implied with `take_while` but we could get side effects if this ever changes.

klochowicz replied:
indeed, perhaps a simple `filter` might be better here as it doesn't need such assumption.

da-kami replied:
Once https://github.com/itchysats/itchysats/pull/2478 merges it should be guaranteed. I will see to fix and merge that one.

luckysori replied:
It's a good point, but I don't see any reason why it shouldn't be sorted. We just append to it during rollover.

It would be nice if could just call `sort_by` on the iterator, but we need the `settlement_id`, which is optional for backwards compatibility.

luckysori committed:
Build new RevokedCommit based on commit TXID proposed by taker

The taker already did so because they could simply _always_ take the
information from the latest DLC.

The maker was supposed to do so, but they were naively using the
information from the latest DLC. They were basically ignoring the
`from_commit_txid` field in the message as soon as they had confirmed
that they could find it in their commit transaction records.

We also had to fix the rollover retry test, which used to pass even
though this bug existed:

On `testnet`, when the maker is ahead with rollovers we run into error
`Counterparty sent invalid revocation sk` when rolling over from a
previous commit. This was not detected by the
`retry_rollover_an_open_cfd` test because, even though we trigger
rolling over from a previous commit-txid, the taker's DLC actually
matched the one of the maker. The test setup did no properly simulate
that the maker is in fact one rollover ahead of the taker.

As it is hard to remove events, the easiest way to simulate the
previous rollover on the taker side, is to add another
`RolloverCompleted` event with the old data. This should be sufficient
to simulate that the taker has in fact the old state and the maker is
ahead.

Co-authored-by: Daniel Karzel <daniel.karzel@coblox.tech>
